docs/source/reference-io.rst

```diff
@@ -1,26 +1,171 @@
+.. currentmodule:: trio
+
 I/O in Trio
 ===========
 
+.. _abstract-stream-api:
+
+The abstract Stream API
+-----------------------
+
+Trio provides a set of abstract base classes that define a standard
+interface for unidirectional and bidirectional byte streams.
+
+Why is this useful? Because it lets you write generic protocol
+implementations that can work over arbitrary transports, and easily
+create complex transport configurations. Here's some examples:
+
+* :class:`trio.SocketStream` wraps a raw socket (like a TCP connection
+  over the network), and converts it to the standard stream interface.
+
+* :class:`trio.ssl.SSLStream` is a "stream adapter" that can take any
+  object that implements the :class:`trio.abc.Stream` interface, and
+  convert it into an encrypted stream. In trio the standard way to
+  speak SSL over the network is to wrap an
+  :class:`~trio.ssl.SSLStream` around a :class:`~trio.SocketStream`.
+
+* If you spawn a subprocess then you can get a
+  :class:`~trio.abc.SendStream` that lets you write to its stdin, and
+  a :class:`~trio.abc.ReceiveStream` that lets you read from its
+  stdout. If for some reason you wanted to speak SSL to a subprocess,
+  you could use a :class:`StapledStream` to combine its stdin/stdout
+  into a single bidirectional :class:`~trio.abc.Stream`, and then wrap
+  that in an :class:`~trio.ssl.SSLStream`::
+
+     ssl_context = trio.ssl.create_default_context()
+     ssl_context.check_hostname = False
+     s = SSLStream(StapledStream(process.stdin, process.stdout), ssl_context)
+
+  [Note: subprocess support is not implemented yet, but that's the
+  plan. Unless it is implemented, and I forgot to remove this note.]
+
+* It sometimes happens that you want to connect to an HTTPS server,
+  but you have to go through a web proxy... and the proxy also uses
+  HTTPS. So you end up having to do `SSL-on-top-of-SSL
+  <https://daniel.haxx.se/blog/2016/11/26/https-proxy-with-curl/>`__. In
+  trio this is trivial – just wrap your first
+  :class:`~trio.ssl.SSLStream` in a second
+  :class:`~trio.ssl.SSLStream`::
+
+     # Get a raw SocketStream connection to the proxy:
+     s0 = await open_tcp_stream("proxy", 443)
```

Comment:
Yeah, that's correct, but #145 will eventually add it I think :-).

In general this PR has gotten kind of out of hand, so I want to try and get it merged and turn most of these comments into follow-up issues. And the docs in particular are in a bit of a mess because there isn't really a good way to document a library that has *half* of a high-level networking API implemented... best thing is to finish it and then figure out how to make the docs good :-). This is a useful reminder to add some sort of "please excuse our dust" note to the docs in the interim though...